api/v1alpha4/maasmachine_webhook.go

```diff
@@ -0,0 +1,75 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var maasmachinelog = logf.Log.WithName("maasmachine-resource")
+
+func (r *MaasMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha4-maasmachine,mutating=true,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=maasmachines,verbs=create;update,versions=v1alpha4,name=mmaasmachine.kb.io,sideEffects=None,admissionReviewVersions=v1beta1;v1
+
+var _ webhook.Defaulter = &MaasMachine{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *MaasMachine) Default() {
+	maasmachinelog.Info("default", "name", r.Name)
+
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha4-maasmachine,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=maasmachines,versions=v1alpha4,name=vmaasmachine.kb.io,sideEffects=None,admissionReviewVersions=v1beta1;v1
+
+var _ webhook.Validator = &MaasMachine{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *MaasMachine) ValidateCreate() error {
+	maasmachinelog.Info("validate create", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object creation.
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *MaasMachine) ValidateUpdate(old runtime.Object) error {
+	maasmachinelog.Info("validate update", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
```

Comment:
please add validation for immutible fields.

Reply:
Made image, minCPU and minMemory immutable for Machine and MachineTemplate